sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -984,6 +984,97 @@ case class GpuFastSampleExec(
   }
 }
 
+class GpuRangeIterator(
+    partitionStart: BigInt,
+    partitionEnd: BigInt,
+    step: Long,
+    maxRowCountPerBatch: Long,
+    taskContext: TaskContext,
+    opTime: GpuMetric) extends Iterator[ColumnarBatch] with Logging {
+
+  private def getSafeMargin(bi: BigInt): Long = {
+    if (bi.isValidLong) {
+      bi.toLong
+    } else if (bi > 0) {
+      Long.MaxValue
+    } else {
+      Long.MinValue
+    }
+  }
+
+  private val safePartitionStart = getSafeMargin(partitionStart) // inclusive
+  private val safePartitionEnd = getSafeMargin(partitionEnd) // exclusive, unless start == this
+  private[this] var number: Long = safePartitionStart
+  private[this] var done: Boolean = false
+
+  override def hasNext: Boolean = {
+    if (!done) {
+      if (step > 0) {
+        number < safePartitionEnd
+      } else {
+        number > safePartitionEnd
+      }
+    } else false
+  }
+
+  override def next(): ColumnarBatch = {
+    GpuSemaphore.acquireIfNecessary(taskContext)
+    withResource(new NvtxWithMetrics("GpuRange", NvtxColor.DARK_GREEN, opTime)) { _ =>
+      val start = number
+      val remainingRows = (safePartitionEnd - start) / step
+      // Start is inclusive so we need to produce at least one row
+      val rowsExpected = Math.max(1, Math.min(remainingRows, maxRowCountPerBatch))
+      val iter = withRetry(AutoCloseableLong(rowsExpected), splitRowsNumberInHalf) { rows =>
+        withResource(Scalar.fromLong(start)) { startScalar =>
+          withResource(Scalar.fromLong(step)) { stepScalar =>
+            withResource(
+                cudf.ColumnVector.sequence(startScalar, stepScalar, rows.value.toInt)) { vec =>
+              withResource(new Table(vec)) { tab =>
+                GpuColumnVector.from(tab, Array[DataType](LongType))
+              }
+            }
+          }
+        }
+      }
+      assert(iter.hasNext)
+      closeOnExcept(iter.next()) { batch =>
+        // it should has only one batch
+        assert(iter.isEmpty)
```

Comment:
This is really confusing because you are doing it differently from everything else that we have done with split and retry. Could you at least document very clearly what is happening both here and in `splitRowsNumberInHalf` so it is clear why there should be one and only one batch, and that is okay because the state that keeps track of this is updated externally.

Reply:
I added more comments. Hope it can help understand better what is happening here.

Yes, it is a little different from the normal way.

Luckily we can do this here, I mean, return a Seq of a single element from the split function. Because the input is NOT a already loaded big batch, but a long value indicating the rows number to be generated. We need multiple elements in the iterator to cache all the rows for the input batch after splitting, but we don't for a single long.

By doing this, we can pull out the single batch directly, and leverage the existing state variables to handle the retry result. If we returns more than one batch by the iterator, we need to maintain the iterator for later loops, leading to more code change. (I think it is not a good idea either to pull out all the batches and concatenate them to return).